File: backend/bot/views/character_views/getCharacterCount.py
from rest_framework.response import Response
from rest_framework import status
from django.views.decorators.csrf import csrf_exempt
from haven.models import Character
from django.db import models
from rest_framework.views import APIView
from ..Authenticate import authenticate


class GetCharacterCountView(APIView):
    """
    API view for retrieving character counts for a user.

    Returns global counts (all characters for the user) and optionally
    guild-specific counts if guild_id is provided.

    Expected POST data:
    - user_id: The ID of the user (required)
    - guild_id: The ID of the guild/chronicle (optional)

    Returns:
    {
        "global": {
            "sheets": int,      // Total character sheets
            "trackers": int,    // Total character trackers
            "total": int        // Total characters (sheets + trackers)
        },
        "guild": {             // Only if guild_id provided
            "sheets": int,      // Guild character sheets
            "trackers": int,    // Guild character trackers
            "total": int        // Guild total characters
        }
    }
    """

    def post(self, request):
        authenticate(request)

        user_id = request.data.get("user_id")
        guild_id = request.data.get("guild_id")

        if not user_id:
            return Response(
                {"error": "user_id is required"}, status=status.HTTP_400_BAD_REQUEST
            )

        # Get global character counts for the user in one query
        global_counts = Character.objects.filter(user_id=user_id)
        global_grouped = global_counts.values("is_sheet").annotate(
            count=models.Count("id")
        )

        global_map = {g["is_sheet"]: g["count"] for g in global_grouped}
        global_sheets = global_map.get(True, 0)
        global_trackers = global_map.get(False, 0)
        global_total = global_sheets + global_trackers
        result = {
            "global": {
                "sheets": global_map.get(True, 0),
                "trackers": global_map.get(False, 0),
                "total": global_total,
            }
        }

        # If guild_id provided, also get guild-specific counts
        if guild_id:
            guild_counts = Character.objects.filter(
                user_id=user_id, chronicle_id=guild_id
            )
            guild_grouped = guild_counts.values("is_sheet").annotate(
                count=models.Count("id")
            )
            guild_map = {g["is_sheet"]: g["count"] for g in guild_grouped}
            guild_sheets = guild_map.get(True, 0)
            guild_trackers = guild_map.get(False, 0)
            guild_total = guild_sheets + guild_trackers
            result["chronicle"] = {
                "sheets": guild_sheets,
                "trackers": guild_trackers,
                "total": guild_total,
            }

        return Response(result, status=status.HTTP_200_OK)

File: discord_bots/src/realm_api/getCharacterCount.js
"use strict";
require(`${process.cwd()}/alias`);
const { postData } = require("./postData.js");
const { RealmError, RealmAPIError, ErrorCodes } = require("@errors");

/**
 * Gets character counts for a user (global and guild-specific)
 * @param {String} user_id - Discord user ID
 * @param {String} guild_id - Discord guild/chronicle ID (optional)
 * @returns {Promise<Object>} Returns character count data
 *
 * @example
 * // Get global character counts
 * const counts = await getCharacterCount("123456789012345678");
 * console.log(counts.global.sheets); // Total sheets across all guilds
 * console.log(counts.global.trackers); // Total trackers across all guilds
 * console.log(counts.global.total); // Total characters across all guilds
 *
 * @example
 * // Get guild-specific character counts
 * const counts = await getCharacterCount(
 *   "123456789012345678",
 *   "987654321098765432"
 * );
 * console.log(counts.global.sheets); // Global counts
 * console.log(counts.guild.sheets); // Guild-specific counts
 * console.log(counts.guild.trackers); // Guild-specific tracker counts
 * console.log(counts.guild.total); // Guild-specific total
 */
module.exports = async function getCharacterCount(userId, guildId) {
  const path = "character/count";

  // Validate required parameters
  if (!userId) {
    throw new RealmError({ cause: new Error("userId must be provided") });
  }

  const data = {
    user_id: userId,
    guild_id: guildId,
  };

  const res = await postData(path, data);

  switch (res?.status) {
    case 200: // Found character counts
      return res.data;
    case 400: // Bad request
      throw new RealmError({ cause: new Error("Invalid parameters provided") });
    default:
      throw new RealmAPIError({
        cause: `res_status: ${res?.status}\nres: ${JSON.stringify(
          res?.data,
          null,
          2
        )}\npost: ${JSON.stringify(data, null, 2)}`,
      });
  }
};
